### ISSUE TITLE
Duplexer Redux solution not working

### ISSUE THREAD

**Issue by Eschon** [workshopper/stream-adventure#158]
I couldn't get duplexer redux to work so I decided to look at the solution. I couldn't find any obvious differences between my program and the solution, so I just copied the solution to see if it works and it doesn't.

Here's the error output:

```
Administrators-Mac-mini:streams egon$ stream-adventure verify duplexer-redux.js

TAP version 13
# (anonymous)
ok 1 stream-adventure verify YOURFILE.js
#########################################
###   YOUR SOLUTION IS NOT CORRECT!   ###
#########################################


ok 2 solution exports a function

events.js:85
      throw er; // Unhandled 'error' event
            ^
TypeError: Invalid non-string/buffer chunk
    at validChunk (/Users/egon/src/nodeschool/streams/node_modules/duplexer2/node_modules/readable-stream/lib/_stream_writable.js:210:14)
    at DuplexWrapper.Writable.write (/Users/egon/src/nodeschool/streams/node_modules/duplexer2/node_modules/readable-stream/lib/_stream_writable.js:237:12)
    at null.<anonymous> (/Users/egon/.nvm/versions/node/v0.12.4/lib/node_modules/stream-adventure/problems/duplexer_redux/index.js:49:20)
    at wrapper [as _onTimeout] (timers.js:265:14)
    at Timer.listOnTimeout (timers.js:110:15)
Administrators-Mac-mini:streams egon$
```

**Comment by atronov** [comment]
+1
I have the same

**Comment by calvinmetcalf** [comment]
[this line](https://github.com/substack/stream-adventure/blob/9ed382598787df13fec063339a1caf1669192ff2/test/solutions/duplexer_redux.js#L1) should likely be `
var duplexer = require('duplexer2').obj;`

**Comment by Eschon** [comment]
I tried to change it and it still doesn't work.
New output:

```
Administrators-Mac-mini:streams egon$ stream-adventure verify duplexer-redux.js

TAP version 13
# (anonymous)
ok 1 stream-adventure verify YOURFILE.js
ok 2 solution exports a function
not ok 3 TypeError: undefined is not a function
  ---
    operator: error
    expected: undefined
    actual:   [TypeError: undefined is not a function]
    stack:
      TypeError: undefined is not a function
        at module.exports (/Users/egon/src/nodeschool/streams/duplexer-redux.js:7:14)
        at /Users/egon/.nvm/versions/node/v0.12.4/lib/node_modules/stream-adventure/problems/duplexer_redux/index.js:39:18
        at Test.<anonymous> (/Users/egon/.nvm/versions/node/v0.12.4/lib/node_modules/stream-adventure/node_modules/adventure-verify/index.js:48:40)
        at Test.bound [as _cb] (/Users/egon/.nvm/versions/node/v0.12.4/lib/node_modules/stream-adventure/node_modules/adventure-verify/node_modules/tape/lib/test.js:60:32)
        at Test.run (/Users/egon/.nvm/versions/node/v0.12.4/lib/node_modules/stream-adventure/node_modules/adventure-verify/node_modules/tape/lib/test.js:74:14)
        at Test.bound [as run] (/Users/egon/.nvm/versions/node/v0.12.4/lib/node_modules/stream-adventure/node_modules/adventure-verify/node_modules/tape/lib/test.js:60:32)
        at Immediate.next [as _onImmediate] (/Users/egon/.nvm/versions/node/v0.12.4/lib/node_modules/stream-adventure/node_modules/adventure-verify/node_modules/tape/lib/results.js:66:15)
        at processImmediate [as _immediateCallback] (timers.js:358:17)
  ...

1..3
# tests 3
# pass  2
# fail  1

#########################################
###   YOUR SOLUTION IS NOT CORRECT!   ###
#########################################
```

**Comment by yuki-scio** [comment]
Hi,

I'm facing this issue, too. I spent hours to tackle on it, however no luck so far. These are my findings so far.

- The solution printed by stream-adventure solution also causes the same error ("Invalid non-string/buffer chunk").
- You need to give `writableObjectMode:true` option to `duplexer()` as [its default is false][option].
   ```
   return duplexer({writableObjectMode:true}, input, counter);
   ```
- I investigated the behavior of the verification code by attaching debugger. It looks that [the handler function of the concat stream][index] isn't executed. I thought the following things should happen:
  - The duplex stream should be flowing mode as it pipes to the concat stream.
  - self.push() calls in counter.setCounts() should trigger 'data' event on the duplex stream
  - The concat stream concatenates all responses pushed by counter.setCounts() and invokes the handler eventually.

I'm stuck at this point. Am I missing something?

[option]: https://nodejs.org/api/stream.html#stream_new_stream_duplex_options
[index]: https://github.com/substack/stream-adventure/blob/master/problems/duplexer_redux/index.js#L41

**Comment by littlejon85** [comment]
Hi,

the solution works after changing duplexer2 -> duplexer and through2 -> through.

The reason for this is beyond me.

**Comment by guidoman** [comment]
If you change:

`return duplexer(input, counter);`

into

`return duplexer({writableObjectMode:true}, input, counter);`

it will work (node 4.2.1, through2 2.0.0, duplexer2 0.1.2).

**Comment by brookinc** [comment]
I was also getting the  `TypeError: Invalid non-string/buffer chunk` issue -- and neither adding `.obj` to the duplexer2 require() nor adding the `{writableObjectMode:true}` block to the return line fixed the issue for me.

However, strangely(?), adding `.obj` to the through2 require() (as per the reference solution) solved the issue for me: `var through = require('through2').obj `

**Comment by benwiley4000** [comment]
thanks @guidoman, adding that option fixed my problem.

@brookinc this is expected behavior, through2 expects buffer chunks unless explicitly told to expect object literals. See the documentation for `through2` here: https://github.com/rvagg/through2

Particularly: "Note that `through2.obj(fn)` is a convenience wrapper around `through2({ objectMode: true }, fn)`." In other words, you could just require through2 and pass a setting to it, similar to what we evidently have to do with `duplexer2`, but `through2.obj` saves you a few characters.

**Comment by brookinc** [comment]
That makes sense. Thanks for the explanation!

**Comment by enderandpeter** [comment]
@guidoman you are absolutely correct. The solution they provided does not say that, so it looks like they should update it.

**Comment by ccarruitero** [comment]
fixed by #172
